### PR TITLE
Optimize for WC linear writes

### PR DESCRIPTION
**Pull request by samikhawaja**
Avoid holes or out of order writes to the command buffers to optimize
write combine bandwidth.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

**Comment by JaxLinAMD** [comment]
ok to test

**Comment by amdvlk-admin** [comment]
### Test summary for commit 98478d326abbb2f9fba05ebbb2da84879c873959
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>LLPC: dev e23176874d33ad107eacb871980b0bfebe8331da</li>
<li>SPVGEN: dev 79f8e32dd7242d7d34fa0074f761c9d794f11811</li>
<li>XGL: dev 5c3c2dd430222cdf1b46c9d385e9c6c5e89604e1</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 0cb96bc82ca13e0fef3684a4ac8f6eff50bce257</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/228128)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31146/57032 (54.6%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 25886/57032 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31145/57032 (54.6%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 25887/57032 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31285/57032 (54.9%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 25747/57032 (45.1%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34853/57032 (61.1%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 22179/57032 (38.9%)</li>
</ul></details>


</ul></details>

File: src/core/hw/gfxip/gfx9/gfx9CmdUtil.cpp [hunk 1]
```diff
@@ -3955,12 +3964,7 @@ size_t CmdUtil::BuildWaitRegMem64(
     pPacket->ordinal8.mask_hi                 = HighPart(mask);
     pPacket->ordinal9.u32All                  = 0;
     pPacket->ordinal9.bitfields.poll_interval = Pal::Device::PollInterval;
-
-    if (Pal::Device::EngineSupportsGraphics(engineType))
-    {
-        pPacket->ordinal2.bitfields.engine_sel = static_cast<ME_WAIT_REG_MEM64_engine_sel_enum>(engine);
-    }
-    else
+    if (!Pal::Device::EngineSupportsGraphics(engineType))
```
Comment (JakeMerdichAMD) [inline src/core/hw/gfxip/gfx9/gfx9CmdUtil.cpp:3967]:
```suggestion
    if (Pal::Device::EngineSupportsGraphics(engineType) == false)
```

**Review by JakeMerdichAMD — CHANGES_REQUESTED** [review]
One code formatting nit, then looks good to me.

**Comment by JakeMerdichAMD** [comment]
Also, this is a very nice find in general, thanks! We tend to be very mindful of RMW over pcie but less so about WC efficiency.

File: src/core/hw/gfxip/gfx9/gfx9CmdUtil.cpp (hunk 1 above)
Comment (samikhawaja) [reply src/core/hw/gfxip/gfx9/gfx9CmdUtil.cpp]: Done

**Comment by amdvlk-admin** [comment]
### Test summary for commit 37012064d32272d111733d46fbf155b45dbe1c54
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>LLPC: dev e23176874d33ad107eacb871980b0bfebe8331da</li>
<li>SPVGEN: dev 79f8e32dd7242d7d34fa0074f761c9d794f11811</li>
<li>XGL: dev 5c3c2dd430222cdf1b46c9d385e9c6c5e89604e1</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 0cb96bc82ca13e0fef3684a4ac8f6eff50bce257</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/228128)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31146/57032 (54.6%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 25886/57032 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31145/57032 (54.6%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 25887/57032 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31285/57032 (54.9%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 25747/57032 (45.1%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34853/57032 (61.1%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 22179/57032 (38.9%)</li>
</ul></details>


</ul></details>

File: src/core/hw/gfxip/gfx9/gfx9CmdUtil.cpp (hunk 1 above)
Comment (samikhawaja) [reply src/core/hw/gfxip/gfx9/gfx9CmdUtil.cpp]: It seems github doesn't like manually doing the suggested change. I made the change you suggested into my branch and force pushed. So it is ready for review again.

**Review by JakeMerdichAMD — APPROVED** [review]

**Pull request merged**